src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -248,11 +249,18 @@ private void OnCompilationEventsGenerated_NoLock(ImmutableArray<CompilationEvent
                         _pooledEventsWithAnyActionsSet.Add(compilationEvent);
 
                         var symbolDeclaredEvent = compilationEvent as SymbolDeclaredCompilationEvent;
-                        if (symbolDeclaredEvent?.DeclaringSyntaxReferences.Length > 1 &&
-                            !_partialSymbolsWithGeneratedSourceEvents.Add(symbolDeclaredEvent.Symbol))
+                        if (symbolDeclaredEvent?.DeclaringSyntaxReferences.Length > 1)
                         {
-                            // already processed.
-                            continue;
+                            if (_partialSymbolsWithGeneratedSourceEvents.Contains(symbolDeclaredEvent.Symbol))
```

Comment:
Concurrency bug here. What happens if two threads do this at the same time (on different parts of the partial). Won't you still get duplicate events?

Isn't this just compensating for a bug in the compiler? Shouldn't we just fix that instead?